File: MercaderSG/Panel/CambiarContrasena.vb
Imports Servicios
Imports Entidades
Imports Negocios
Imports Excepciones
Public Class CambiarContrasena

    Private Sub CambiarContrasena_Load(sender As System.Object, e As System.EventArgs) Handles MyBase.Load
        CargarTT()
        AplicarIdioma()
        CargarFoco(ContrasenaGB)
    End Sub

    Private Sub AceptarBtn_Click(sender As System.Object, e As System.EventArgs) Handles AceptarBtn.Click
        If Not ConsistenciaDatos() Then
            Exit Sub
        End If

        Dim ConAnt As String = ConAnteriorTxt.Text
        Dim NuevaCon As String = NuevaContraseñaTxt.Text
        Dim ReNuevaCon As String = ReNuevaContraseñaTxt.Text
        Dim UsuAut As Autenticar = Autenticar.Instancia()

        If NuevaCon = ReNuevaCon Then
            Try
                UsuarioRN.CambiarContraseña(UsuAut.UsuarioLogueado, ConAnt, NuevaCon)
            Catch ex As InformationException
                MessageBox.Show(ex.Message, My.Resources.ArchivoIdioma.MsgBoxInformacion, MessageBoxButtons.OK, MessageBoxIcon.Information)
                Me.Close()
            Catch ex As WarningException
                MessageBox.Show(ex.Message, My.Resources.ArchivoIdioma.MsgBoxAdvertencia, MessageBoxButtons.OK, MessageBoxIcon.Warning)

                If ex.Message = My.Resources.ArchivoIdioma.UsuarioDadoBaja Then
                    Me.Close()
                Else
                    ConAnteriorTxt.Clear()
                End If
            End Try
        Else
            MessageBox.Show(My.Resources.ArchivoIdioma.NoCoincidenPass, My.Resources.ArchivoIdioma.MsgBoxAdvertencia, MessageBoxButtons.OK, MessageBoxIcon.Warning)
            NuevaContraseñaTxt.Clear()
            ReNuevaContraseñaTxt.Clear()
        End If
    End Sub

    Private Sub CancelarBtn_Click(sender As System.Object, e As System.EventArgs) Handles CancelarBtn.Click
        Me.Close()
    End Sub

#Region "Validaciones"

    Private Function ConsistenciaDatos() As Boolean
        Dim Resultado As Boolean = True

        If String.IsNullOrEmpty(ConAnteriorTxt.Text) And String.IsNullOrEmpty(NuevaContraseñaTxt.Text) And String.IsNullOrEmpty(ReNuevaContraseñaTxt.Text) Then
            MessageBox.Show(My.Resources.ArchivoIdioma.Completar3CamposCon, My.Resources.ArchivoIdioma.MsgBoxAdvertencia, MessageBoxButtons.OK, MessageBoxIcon.Warning)
            ConAnteriorTxt.Focus()
            Resultado = False

            Return Resultado
        End If

        'Contraseña Anterior
        If String.IsNullOrEmpty(ConAnteriorTxt.Text) Then
            MensajeTT.Show(My.Resources.ArchivoIdioma.CampoObligatorio, ConAnteriorTxt)
            ConAnteriorTxt.Focus()
            Resultado = False

            Return Resultado
        End If

        Select Case ConAnteriorTxt.Text.Length
            Case 1 To 7
                MensajeTT.Show(My.Resources.ArchivoIdioma.ContenerMenos8Carac, ConAnteriorTxt)
                ConAnteriorTxt.Focus()
                Resultado = False

                Return Resultado

            Case 8 To 12
                MensajeTT.Hide(ConAnteriorTxt)

            Case Is > 12
                MensajeTT.Show(My.Resources.ArchivoIdioma.Contener12Carac, ConAnteriorTxt)
                ConAnteriorTxt.Focus()
                Resultado = False

                Return Resultado
        End Select

        'Nueva Contraseña

        If String.IsNullOrEmpty(NuevaContraseñaTxt.Text) Then
            MensajeTT.Show(My.Resources.ArchivoIdioma.CampoObligatorio, NuevaContraseñaTxt)
            NuevaContraseñaTxt.Focus()
            Resultado = False

            Return Resultado
        End If

        Select Case NuevaContraseñaTxt.Text.Length
            Case 1 To 7
                MensajeTT.Show(My.Resources.ArchivoIdioma.ContenerMenos8Carac, NuevaContraseñaTxt)
                NuevaContraseñaTxt.Focus()
                Resultado = False

                Return Resultado

            Case 8 To 12
                MensajeTT.Hide(NuevaContraseñaTxt)

            Case Is > 12
                MensajeTT.Show(My.Resources.ArchivoIdioma.Contener12Carac, NuevaContraseñaTxt)
                NuevaContraseñaTxt.Focus()
                Resultado = False

                Return Resultado
        End Select

        'Repetir Contraseña

        If String.IsNullOrEmpty(ReNuevaContraseñaTxt.Text) Then
            MensajeTT.Show(My.Resources.ArchivoIdioma.CampoObligatorio, ReNuevaContraseñaTxt)
            ReNuevaContraseñaTxt.Focus()
            Resultado = False

            Return Resultado
        End If

        Select Case ReNuevaContraseñaTxt.Text.Length
            Case 1 To 7
                MensajeTT.Show(My.Resources.ArchivoIdioma.ContenerMenos8Carac, ReNuevaContraseñaTxt)
                ReNuevaContraseñaTxt.Focus()
                Resultado = False

                Return Resultado

            Case 8 To 12
                MensajeTT.Hide(ReNuevaContraseñaTxt)

            Case Is > 12
                MensajeTT.Show(My.Resources.ArchivoIdioma.Contener12Carac, ReNuevaContraseñaTxt)
                ReNuevaContraseñaTxt.Focus()
                Resultado = False

                Return Resultado
        End Select

        Return Resultado
    End Function


    Private Sub ContraseñasTxts_KeyPress(sender As System.Object, e As System.Windows.Forms.KeyPressEventArgs) Handles ConAnteriorTxt.KeyPress, NuevaContraseñaTxt.KeyPress, ReNuevaContraseñaTxt.KeyPress
        Dim CaracteresPermitidos As String = "qwertyuiopasdfghjklñzxcvbnmQWERTYUIOPASDFGHJKLÑZXCVBNM1234567890~!@#$%^&*()_-+={}[]\|:;<>,.?/" & Convert.ToChar(8)
        Dim c As Char = e.KeyChar
        If Not CaracteresPermitidos.Contains(c) Then
            e.Handled = True
        End If
    End Sub

    Private Sub ConAnteriorTxt_TextChanged(sender As System.Object, e As System.EventArgs) Handles ConAnteriorTxt.TextChanged
        MensajeTT.Hide(ConAnteriorTxt)
    End Sub

    Private Sub NuevaContraseñaTxt_TextChanged(sender As System.Object, e As System.EventArgs) Handles NuevaContraseñaTxt.TextChanged
        MensajeTT.Hide(NuevaContraseñaTxt)
    End Sub

    Private Sub ReNuevaContraseñaTxt_TextChanged(sender As System.Object, e As System.EventArgs) Handles ReNuevaContraseñaTxt.TextChanged
        MensajeTT.Hide(ReNuevaContraseñaTxt)
    End Sub

    Private Sub CambiarContrasena_FormClosing(sender As System.Object, e As System.Windows.Forms.FormClosingEventArgs) Handles MyBase.FormClosing
        Principal.CambiarContrasenaSMI.Enabled = True
    End Sub

    Private Sub CambiarContrasena_KeyDown(sender As System.Object, e As System.Windows.Forms.KeyEventArgs) Handles MyBase.KeyDown
        If e.KeyCode = Keys.Escape Then
            Me.Close()
        End If
    End Sub

#End Region

#Region "Mejoras"

    Private Sub TieneFoco(ByVal sender As System.Object, ByVal e As System.EventArgs)
        Dim MiTextBox As TextBox = DirectCast(sender, TextBox)
        MiTextBox.BackColor = Color.WhiteSmoke
    End Sub

    Private Sub PierdeFoco(ByVal sender As Object, ByVal e As System.EventArgs)
        Dim MiTextBox As TextBox = DirectCast(sender, TextBox)
        MiTextBox.BackColor = Color.White
    End Sub

    Public Sub CargarFoco(Grupo As Windows.Forms.GroupBox)
        Dim Ctrl As Control
        For Each Ctrl In Grupo.Controls
            If (TypeOf (Ctrl) Is System.Windows.Forms.TextBox) Then
                Dim MiTextBox As TextBox = DirectCast(Ctrl, TextBox)
                If MiTextBox.ReadOnly = False Then
                    AddHandler MiTextBox.GotFocus, AddressOf TieneFoco
                    AddHandler MiTextBox.LostFocus, AddressOf PierdeFoco
                End If
            End If
        Next
    End Sub

#End Region

#Region "Idioma y TT"

    Public Sub AplicarIdioma()
        Me.Text = My.Resources.ArchivoIdioma.CambiarContraseñaFrm

        ContrasenaGB.Text = My.Resources.ArchivoIdioma.CambiarContraseña
        ConAnteriorLbl.Text = My.Resources.ArchivoIdioma.ConAnteriorLbl
        NuevaContraseñaLbl.Text = My.Resources.ArchivoIdioma.NuevaConLbl
        ReNuevaContraseñaLbl.Text = My.Resources.ArchivoIdioma.ReNuevaConLbl

        AceptarBtn.Text = My.Resources.ArchivoIdioma.AceptarBtn
        CancelarBtn.Text = My.Resources.ArchivoIdioma.CancelarBtn
    End Sub

    Private Sub CargarTT()
        ControlesTP.SetToolTip(ConAnteriorTxt, My.Resources.ArchivoIdioma.TTConAnterior)
        ControlesTP.SetToolTip(NuevaContraseñaTxt, My.Resources.ArchivoIdioma.TTNuevaCon)
        ControlesTP.SetToolTip(ReNuevaContraseñaTxt, My.Resources.ArchivoIdioma.TTConfirmarCon)
        ControlesTP.SetToolTip(AceptarBtn, My.Resources.ArchivoIdioma.CambiarContraseña)
        ControlesTP.SetToolTip(CancelarBtn, My.Resources.ArchivoIdioma.TTCancelarBtn)
    End Sub

#End Region

End Class

File: MercaderSG/Panel/CambiarContrasena.cs
using System;
using System.Drawing;
using System.Windows.Forms;
using Excepciones;
using Microsoft.VisualBasic.CompilerServices;
using Negocios;
using Servicios;

namespace MercaderSG
{
    public partial class CambiarContrasena
    {
        public CambiarContrasena()
        {
            InitializeComponent();
            _ConAnteriorTxt.Name = "ConAnteriorTxt";
            _NuevaContraseñaTxt.Name = "NuevaContraseñaTxt";
            _ReNuevaContraseñaTxt.Name = "ReNuevaContraseñaTxt";
            _CancelarBtn.Name = "CancelarBtn";
            _AceptarBtn.Name = "AceptarBtn";
        }

        private void CambiarContrasena_Load(object sender, EventArgs e)
        {
            CargarTT();
            AplicarIdioma();
            CargarFoco(ContrasenaGB);
        }

        private void AceptarBtn_Click(object sender, EventArgs e)
        {
            if (!ConsistenciaDatos())
            {
                return;
            }

            string ConAnt = ConAnteriorTxt.Text;
            string NuevaCon = NuevaContraseñaTxt.Text;
            string ReNuevaCon = ReNuevaContraseñaTxt.Text;
            var UsuAut = Autenticar.Instancia();
            if ((NuevaCon ?? "") == (ReNuevaCon ?? ""))
            {
                try
                {
                    UsuarioRN.CambiarContraseña(UsuAut.UsuarioLogueado, ConAnt, NuevaCon);
                }
                catch (InformationException ex)
                {
                    MessageBox.Show(ex.Message, My.Resources.ArchivoIdioma.MsgBoxInformacion, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    Close();
                }
                catch (WarningException ex)
                {
                    MessageBox.Show(ex.Message, My.Resources.ArchivoIdioma.MsgBoxAdvertencia, MessageBoxButtons.OK, MessageBoxIcon.Warning);
                    if ((ex.Message ?? "") == (My.Resources.ArchivoIdioma.UsuarioDadoBaja ?? ""))
                    {
                        Close();
                    }
                    else
                    {
                        ConAnteriorTxt.Clear();
                    }
                }
            }
            else
            {
                MessageBox.Show(My.Resources.ArchivoIdioma.NoCoincidenPass, My.Resources.ArchivoIdioma.MsgBoxAdvertencia, MessageBoxButtons.OK, MessageBoxIcon.Warning);
                NuevaContraseñaTxt.Clear();
                ReNuevaContraseñaTxt.Clear();
            }
        }

        private void CancelarBtn_Click(object sender, EventArgs e)
        {
            Close();
        }

        /* TODO ERROR: Skipped RegionDirectiveTrivia */
        private bool ConsistenciaDatos()
        {
            bool Resultado = true;
            if (string.IsNullOrEmpty(ConAnteriorTxt.Text) & string.IsNullOrEmpty(NuevaContraseñaTxt.Text) & string.IsNullOrEmpty(ReNuevaContraseñaTxt.Text))
            {
                MessageBox.Show(My.Resources.ArchivoIdioma.Completar3CamposCon, My.Resources.ArchivoIdioma.MsgBoxAdvertencia, MessageBoxButtons.OK, MessageBoxIcon.Warning);
                ConAnteriorTxt.Focus();
                Resultado = false;
                return Resultado;
            }

            // Contraseña Anterior
            if (string.IsNullOrEmpty(ConAnteriorTxt.Text))
            {
                MensajeTT.Show(My.Resources.ArchivoIdioma.CampoObligatorio, ConAnteriorTxt);
                ConAnteriorTxt.Focus();
                Resultado = false;
                return Resultado;
            }

            switch (ConAnteriorTxt.Text.Length)
            {
                case var @case when 1 <= @case && @case <= 7:
                    {
                        MensajeTT.Show(My.Resources.ArchivoIdioma.ContenerMenos8Carac, ConAnteriorTxt);
                        ConAnteriorTxt.Focus();
                        Resultado = false;
                        return Resultado;
                    }

                case var case1 when 8 <= case1 && case1 <= 12:
                    {
                        MensajeTT.Hide(ConAnteriorTxt);
                        break;
                    }

                case var case2 when case2 > 12:
                    {
                        MensajeTT.Show(My.Resources.ArchivoIdioma.Contener12Carac, ConAnteriorTxt);
                        ConAnteriorTxt.Focus();
                        Resultado = false;
                        return Resultado;
                    }
            }

            // Nueva Contraseña

            if (string.IsNullOrEmpty(NuevaContraseñaTxt.Text))
            {
                MensajeTT.Show(My.Resources.ArchivoIdioma.CampoObligatorio, NuevaContraseñaTxt);
                NuevaContraseñaTxt.Focus();
                Resultado = false;
                return Resultado;
            }

            switch (NuevaContraseñaTxt.Text.Length)
            {
                case var case3 when 1 <= case3 && case3 <= 7:
                    {
                        MensajeTT.Show(My.Resources.ArchivoIdioma.ContenerMenos8Carac, NuevaContraseñaTxt);
                        NuevaContraseñaTxt.Focus();
                        Resultado = false;
                        return Resultado;
                    }

                case var case4 when 8 <= case4 && case4 <= 12:
                    {
                        MensajeTT.Hide(NuevaContraseñaTxt);
                        break;
                    }

                case var case5 when case5 > 12:
                    {
                        MensajeTT.Show(My.Resources.ArchivoIdioma.Contener12Carac, NuevaContraseñaTxt);
                        NuevaContraseñaTxt.Focus();
                        Resultado = false;
                        return Resultado;
                    }
            }

            // Repetir Contraseña

            if (string.IsNullOrEmpty(ReNuevaContraseñaTxt.Text))
            {
                MensajeTT.Show(My.Resources.ArchivoIdioma.CampoObligatorio, ReNuevaContraseñaTxt);
                ReNuevaContraseñaTxt.Focus();
                Resultado = false;
                return Resultado;
            }

            switch (ReNuevaContraseñaTxt.Text.Length)
            {
                case var case6 when 1 <= case6 && case6 <= 7:
                    {
                        MensajeTT.Show(My.Resources.ArchivoIdioma.ContenerMenos8Carac, ReNuevaContraseñaTxt);
                        ReNuevaContraseñaTxt.Focus();
                        Resultado = false;
                        return Resultado;
                    }

                case var case7 when 8 <= case7 && case7 <= 12:
                    {
                        MensajeTT.Hide(ReNuevaContraseñaTxt);
                        break;
                    }

                case var case8 when case8 > 12:
                    {
                        MensajeTT.Show(My.Resources.ArchivoIdioma.Contener12Carac, ReNuevaContraseñaTxt);
                        ReNuevaContraseñaTxt.Focus();
                        Resultado = false;
                        return Resultado;
                    }
            }

            return Resultado;
        }

        private void ContraseñasTxts_KeyPress(object sender, KeyPressEventArgs e)
        {
            string CaracteresPermitidos = @"qwertyuiopasdfghjklñzxcvbnmQWERTYUIOPASDFGHJKLÑZXCVBNM1234567890~!@#$%^&*()_-+={}[]\|:;<>,.?/" + Convert.ToChar(8);
            char c = e.KeyChar;
            if (!CaracteresPermitidos.Contains(Conversions.ToString(c)))
            {
                e.Handled = true;
            }
        }

        private void ConAnteriorTxt_TextChanged(object sender, EventArgs e)
        {
            MensajeTT.Hide(ConAnteriorTxt);
        }

        private void NuevaContraseñaTxt_TextChanged(object sender, EventArgs e)
        {
            MensajeTT.Hide(NuevaContraseñaTxt);
        }

        private void ReNuevaContraseñaTxt_TextChanged(object sender, EventArgs e)
        {
            MensajeTT.Hide(ReNuevaContraseñaTxt);
        }

        private void CambiarContrasena_FormClosing(object sender, FormClosingEventArgs e)
        {
            My.MyProject.Forms.Principal.CambiarContrasenaSMI.Enabled = true;
        }

        private void CambiarContrasena_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyCode == Keys.Escape)
            {
                Close();
            }
        }

        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
        /* TODO ERROR: Skipped RegionDirectiveTrivia */
        private void TieneFoco(object sender, EventArgs e)
        {
            TextBox MiTextBox = (TextBox)sender;
            MiTextBox.BackColor = Color.WhiteSmoke;
        }

        private void PierdeFoco(object sender, EventArgs e)
        {
            TextBox MiTextBox = (TextBox)sender;
            MiTextBox.BackColor = Color.White;
        }

        public void CargarFoco(GroupBox Grupo)
        {
            foreach (Control Ctrl in Grupo.Controls)
            {
                if (Ctrl is TextBox)
                {
                    TextBox MiTextBox = (TextBox)Ctrl;
                    if (MiTextBox.ReadOnly == false)
                    {
                        MiTextBox.GotFocus += TieneFoco;
                        MiTextBox.LostFocus += PierdeFoco;
                    }
                }
            }
        }

        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
        /* TODO ERROR: Skipped RegionDirectiveTrivia */
        public void AplicarIdioma()
        {
            Text = My.Resources.ArchivoIdioma.CambiarContraseñaFrm;
            ContrasenaGB.Text = My.Resources.ArchivoIdioma.CambiarContraseña;
            ConAnteriorLbl.Text = My.Resources.ArchivoIdioma.ConAnteriorLbl;
            NuevaContraseñaLbl.Text = My.Resources.ArchivoIdioma.NuevaConLbl;
            ReNuevaContraseñaLbl.Text = My.Resources.ArchivoIdioma.ReNuevaConLbl;
            AceptarBtn.Text = My.Resources.ArchivoIdioma.AceptarBtn;
            CancelarBtn.Text = My.Resources.ArchivoIdioma.CancelarBtn;
        }

        private void CargarTT()
        {
            ControlesTP.SetToolTip(ConAnteriorTxt, My.Resources.ArchivoIdioma.TTConAnterior);
            ControlesTP.SetToolTip(NuevaContraseñaTxt, My.Resources.ArchivoIdioma.TTNuevaCon);
            ControlesTP.SetToolTip(ReNuevaContraseñaTxt, My.Resources.ArchivoIdioma.TTConfirmarCon);
            ControlesTP.SetToolTip(AceptarBtn, My.Resources.ArchivoIdioma.CambiarContraseña);
            ControlesTP.SetToolTip(CancelarBtn, My.Resources.ArchivoIdioma.TTCancelarBtn);
        }

        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
    }
}